### DISCUSSION TITLE
Understanding "tune" optimization algorithm

### DISCUSSION THREAD

**Discussion by KenleyData** [kogalur/randomForestSRC#199] [Q&A] [answered]
I recently used the "tune" function to optimize hyperparameters, and as I'm looking at the documentation and my results I'm confused about how the algorithm works. The documentation says that the "stepFactor" input would determine the steps in the parameter "mtry". I set "mtryStart = 5" and "stepFactor=2", but the values of "mtry" that were tested were 1, 2, 3, 5, 10, 15, and 20. I am confused that smaller and bigger steps were taken. These results are fine for my purposes, but I would like to figure out what I am misunderstanding about the algorithm.

Thanks very much for any insight into how the algorithm optimizes the "mtry" variable. Code below:

tune(formula, data = early_cohort, mtryStart = 5, nodesizeTry = c(5,15,25), ntreeTry=100, stepFactor=2,trace=TRUE)

**Comment by ishwaran** [comment, marked as answer]
The algorithm uses a bisection method when searching `mtry`.  The value `mtryStart` is used to initialize the starting value and the algorithm moves left or right by scaling the value of `mtry` by `stepFactor`.  With a factor of 2, the value is increased by a factor of 2 when moving to the right and decreased by a factor of 2 when moving to the left.  Values are rounded either up or down to integers.

**Reply by KenleyData** [reply]
Thank you, this is very helpful! Could you please clarify how the algorithm decides to move left or right in seeking the optimal value of mtry? (Please let me know if there is documentation that covers these questions so that I am not spamming the board with questions that are already documented.)

**Reply by KenleyData** [reply]
Also to clarify--by "left" you mean lower numbers and "right" you mean higher numbers?

**Reply by ishwaran** [reply]
It searches over both directions (left meaning lower `mtry` values, right meaning higher `mtry` values) in each loop, so when it goes left, it divides by 0.5, or when it goes right, it multiplies by 2, if the scale factor is 2.

I suggest you simply output the function to the terminal (or save it as a file) if you have further questions. It's a relatively simple function and its easier to explain if you simply look at it.

This will output the function for you to the terminal:
> tune.rfsrc

**Reply by KenleyData** [reply]
Got it, I read through the code and I think I understand now. Thanks!
